File: algorithms/medium/car_pooling.ex
defmodule Solution do
  @spec car_pooling(trips :: [[integer]], capacity :: integer) :: boolean
  def car_pooling(trips, capacity) do
    m =
      trips
      |> Enum.reduce(%{}, fn [n, from, to], acc ->
        acc
        |> Map.update(from, n, &(&1 + n))
        |> Map.update(to, -n, &(&1 - n))
      end)

    0..1000
    |> Enum.reduce_while({capacity, true}, fn i, {capacity, res} ->
      n = Map.get(m, i, 0)

      if capacity < n do
        {:halt, {capacity, false}}
      else
        {:cont, {capacity - n, true}}
      end
    end)
    |> elem(1)
  end
end

File: algorithms/medium/car_pooling.rs
impl Solution {
    pub fn car_pooling(trips: Vec<Vec<i32>>, mut capacity: i32) -> bool {
        let mut arr = vec![0; 1001];

        for trip in &trips {
            if let &[n, from, to] = &trip[..3] {
                arr[from as usize] += n;
                arr[to as usize] -= n;
            }
        }

        for &n in &arr {
            if capacity < n {
                return false;
            }
            capacity -= n;
        }

        true
    }
}
